File: dynamic_programming/overlapping_subproblems.py
# http://www.geeksforgeeks.org/dynamic-programming-set-1/

# MEMOIZATION
def fibonacci(n, lookup):
  # Base case
  if n < 2:
    lookup[n] = n

  # If the value is not calculated previously 
  # then calculate it
  if lookup[n] is None:
    lookup[n] = fibonacci(n-1, lookup) + fibonacci(n-2, lookup)

  return lookup[n]

print('Memoization: top down method')
print('Fibonacci Number is ', fibonacci(34, [None]*(101)))





# TABULATION
def fibonacci2(n):
  # Declare size of array
  f = [0]*(n+1)

  # Base declaration
  f[1] = 1

  for i in range(2, n+1):
    f[i] = f[i-1] + f[i-2]
  return f[n]

print('Tabulation: bottom up method')
print('Fibonacci Number is ', fibonacci(34))

File: dynamic_programming/overlapping_subproblems.js
// MEMOIZATION
// Method 1, pre-defined lookup table
function fibonacci(n, lookup) {
  if(n < 2) {
    lookup[n] = n;
  };

  if(n >= lookup.length) {
    lookup[n] = fibonacci(n-1, lookup) + fibonacci(n-2, lookup);
  };

  return lookup[n];
};

fibonacci(34, []);


// Method 2, user don't need to specifiy lookup
function fibonacci(n, lookup=[]) {

  if(n < 2) {
    lookup[n] = n;
  };

  if(n >= lookup.length) {
    lookup[n] = fibonacci(n-1, lookup) + fibonacci(n-2, lookup);
  };

  return lookup[n];
};

fibonacci(34);


// Method 3, don't build lookup table from empty array
function fibonacci(n, lookup=[0,1]) {
  if(n >= lookup.length) {
    lookup[n] = fibonacci(n-1, lookup) + fibonacci(n-2, lookup);
  };

  return lookup[n];
}

fibonacci(34);





// TABULATION
function fibonacci(n) {
  var lookup = [0,1];

  var i = 2;
  for(i; i < n+1; i++) {
    lookup[i] = lookup[i-1] + lookup[i-2];
  };

  return lookup[n];
};

fibonacci(34);